CMakeLists.txt

```diff
@@ -4,29 +4,131 @@
 
 cmake_minimum_required(VERSION 2.8)
 
-if(DEFINED CMAKE_BUILD_TYPE)
-   set(CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE} CACHE STRING "Debug Release RelWithDebInfo MinSizeRel")
-else()
-   set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Debug Release RelWithDebInfo MinSizeRel")
+project(uavcan C CXX)
+
+#=============================================================================
+# build options
+#
+# options are listed in a table format below
+set(opts
+    # name: type: default value: string options list : description
+    "CMAKE_BUILD_TYPE:STRING:RelWithDebInfo:Debug Release RelWithDebInfo MinSizeRel:Build type."
+    "CMAKE_CXX_FLAGS:STRING:::C++ flags."
+    "CMAKE_C_FLAGS:STRING:::C flags."
+    "UAVCAN_USE_CPP03:BOOL:OFF::Use cpp03 standard."
+    "UAVCAN_NO_ASSERTIONS:BOOL:OFF::Enable assertions."
+    "UAVCAN_USE_EXTERNAL_SNPRINTF:BOOL:OFF::Use external snprintf."
+    "UAVCAN_MEM_POOL_BLOCK_SIZE:STRING:48::memory pool block size"
+    "UAVCAN_STM32_TIMER_NUMBER:STRING:5::default timer"
+    "UAVCAN_MAX_NETWORK_SIZE_HINT:STRING:16::max network size"
+    "UAVCAN_PLATFORM:STRING:generic:generic linux stm32:Platform."
+    "UAVCAN_OS:STRING:generic:generic posix nuttx chibios:Operating system."
+    "UAVCAN_STM32_NUM_IFACES:STRING:2:1 2:Number of interfaces."
+    )
```

Comment:
Is it really necessary to duplicate options here? This violates DRY and requires extra maintenance. I suggest to keep using them via preprocessor, like we do with make.